lib/internal/modules/esm/worker.js

```diff
@@ -0,0 +1,208 @@
+    }
+    unsettledResponsePorts.clear();
+
+    AtomicsAdd(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 1);
+    AtomicsNotify(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION);
+
+    // Let's attach back the event handler.
+    syncCommPort.on('message', handleMessage);
+    // Let's also check synchronously for a message, in case it's already there.
+    clearImmediate(immediate);
+    checkForMessages();
+    // We don't need the sync check after this tick, as we already have added the event handler.
+    clearImmediate(immediate);
+    // Let's add some work for next tick so the worker cannot exit.
+    setImmediate(() => {});
+  });
+
+  async function handleMessage({ method, args, port }) {
+    // Each potential exception needs to be caught individually so that the correct error is sent to
+    // the main thread.
+    let hasError = false;
+    let shouldRemoveGlobalErrorHandler = false;
+    assert(typeof hooks[method] === 'function');
+    if (port == null && !hasUncaughtExceptionCaptureCallback()) {
+      // When receiving sync messages, we want to unlock the main thread when there's an exception.
+      process.on('uncaughtException', errorHandler);
+      shouldRemoveGlobalErrorHandler = true;
+    }
+
+    // We are about to yield the execution with `await ReflectApply` below. In case the code
+    // following the `await` never runs, we remove the message handler so the `beforeExit` event
+    // can be triggered.
+    syncCommPort.off('message', handleMessage);
+
+    // We keep checking for new messages to not miss any.
+    clearImmediate(immediate);
+    immediate = setImmediate(checkForMessages).unref();
+
+    unsettledResponsePorts.add(port ?? syncCommPort);
+
+    let response;
+    try {
+      response = await ReflectApply(hooks[method], hooks, args);
+    } catch (exception) {
+      hasError = true;
+      response = exception;
+    }
+
+    unsettledResponsePorts.delete(port ?? syncCommPort);
+
+    // Send the method response (or exception) to the main thread.
+    try {
+      (port ?? syncCommPort).postMessage(
+        wrapMessage(hasError ? 'error' : 'success', response),
+        transferArrayBuffer(hasError, response?.source),
+      );
+    } catch (exception) {
+      // Or send the exception thrown when trying to send the response.
+      (port ?? syncCommPort).postMessage(wrapMessage('error', exception));
+    }
+
+    AtomicsAdd(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 1);
+    AtomicsNotify(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION);
+    if (shouldRemoveGlobalErrorHandler) {
+      process.off('uncaughtException', errorHandler);
+    }
+
+    syncCommPort.off('message', handleMessage);
+    // We keep checking for new messages to not miss any.
+    clearImmediate(immediate);
+    immediate = setImmediate(checkForMessages).unref();
+  }
+}
+
+/**
+ * ! Run everything possible within this function so errors get reported.
+ */
+module.exports = function(workerData, syncCommPort) {
```

Comment:
for the stack-trace:

```suggestion
module.exports = function setupModuleWorker(workerData, syncCommPort) {
```